instrumentation/gopkg.in/olivere/elastic/splunkelastic/elastic.go

```diff
@@ -0,0 +1,157 @@
+	cfg.DefaultStartOpts = append([]trace.SpanStartOption{
+		trace.WithSpanKind(trace.SpanKindClient),
+		trace.WithAttributes(semconv.DBSystemElasticsearch),
+	}, cfg.DefaultStartOpts...)
+
+	return &roundTripper{RoundTripper: rt, cfg: cfg}
+}
+
+// roundTripper wraps an http.RoundTripper's requests with a span.
+type roundTripper struct {
+	http.RoundTripper
+
+	cfg *internal.Config
+}
+
+var _ http.RoundTripper = (*roundTripper)(nil)
+
+func (rt *roundTripper) RoundTrip(r *http.Request) (resp *http.Response, err error) {
+	opts := rt.cfg.MergedSpanStartOptions(
+		trace.WithAttributes(semconv.HTTPClientAttributesFromHTTPRequest(r)...),
+		trace.WithAttributes(semconv.NetAttributesFromHTTPRequest("tcp", r)...),
+	)
+
+	tracer := rt.cfg.ResolveTracer(r.Context())
+	ctx, span := tracer.Start(r.Context(), name(r), opts...)
+
+	// Ensure anything downstream knows about the started span.
+	r = r.WithContext(ctx)
+	rt.cfg.Propagator.Inject(ctx, propagation.HeaderCarrier(r.Header))
+
+	resp, err = rt.RoundTripper.RoundTrip(r)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		span.End()
+		return
+	}
+
+	span.SetAttributes(semconv.HTTPAttributesFromHTTPStatusCode(resp.StatusCode)...)
+	span.SetStatus(semconv.SpanStatusFromHTTPStatusCode(resp.StatusCode))
+	resp.Body = &wrappedBody{ctx: ctx, span: span, body: resp.Body}
+
+	return
+}
+
+// name returns an appropriate span name based on the client request.
+// OpenTelemetry semantic conventions require this name to be low cardinality,
+// but since the Elasticsearch API is somewhat predictable we can usually
+// return more than just "HTTP {METHOD}". If this is a recognized
+// Elasticsearch operation the returned span name will conform with
+// OpenTelemetry database semantics, otherwise HTTP semantics will be used.
+func name(r *http.Request) string {
+	path := r.URL.Path
+	if path == "" {
+		path = "/"
+	}
+
+	tokenized := tokenize(path)
+	if tokenized == "" {
+		// Unrecognized Elasticsearch path, default to HTTP semantics.
+		return "HTTP " + r.Method
+	}
+
+	op, ok := operations[url{method: r.Method, path: tokenized}]
+	if !ok {
+		// Unrecognized Elasticsearch operation, default to HTTP semantics.
+		return "HTTP " + r.Method + " " + tokenized
+	}
+
+	if strings.HasPrefix(tokenized, "/{index}") {
+		// Use the {index} as the OpenTelemetry semantic for DB name.
+		p := strings.TrimPrefix(path, "/")
+		// Either: ["example-index"] or ["example-index", "*"]
+		const nParts = 2
+		idx := strings.SplitN(p, "/", nParts)[0]
+		// <db.operation> <db.name>
+		return op + " " + idx
+	}
+
+	// <db.operation>
+	return op
+}
+
+type wrappedBody struct {
+	ctx  context.Context
+	span trace.Span
+	body io.ReadCloser
+}
+
+var _ io.ReadCloser = (*wrappedBody)(nil)
+
+func (wb *wrappedBody) Read(b []byte) (int, error) {
+	n, err := wb.body.Read(b)
+	switch err {
+	case nil:
+		// nothing to do here but fall through to the return
+	case io.EOF:
+		wb.span.End()
```

Comment:
Can this happen and Close() be called on the same body object? would span handle double End() calls gracefully?

Reply:
It will be implementation dependent, but the default SDK this distro uses does handle multiple calls to `End` gracefully (in that the subsequent calls are ignored).